hash_reader.go

```diff
@@ -58,228 +107,277 @@ func NewHashReader(ionReader ion.Reader, hasherProvider IonHasherProvider) (Hash
 	return &hashReader{ionReader: ionReader, hasher: *newHasher}, nil
 }
 
-func (hashReader *hashReader) SymbolTable() ion.SymbolTable {
-	return hashReader.ionReader.SymbolTable()
+// SymbolTable returns the current symbol table, or nil if there isn't one.
+// Text Readers do not, generally speaking, have an associated symbol table.
+// Binary Readers do.
+func (hr *hashReader) SymbolTable() ion.SymbolTable {
+	return hr.ionReader.SymbolTable()
 }
 
-func (hashReader *hashReader) Next() bool {
-	hashReader.err = nil
+// Next advances the Reader to the next position in the current value stream.
+// It returns true if this is the position of an Ion value, and false if it
+// is not. On error, it returns false and sets Err.
+func (hr *hashReader) Next() bool {
+	hr.err = nil
 
-	if hashReader.currentType != ion.NoType {
-		if ion.IsScalar(hashReader.currentType) || hashReader.IsNull() {
-			hashReader.err = hashReader.hasher.scalar(hashReader)
-			if hashReader.err != nil {
+	if hr.currentType != ion.NoType {
+		if ion.IsScalar(hr.currentType) || hr.IsNull() {
+			hr.err = hr.hasher.scalar(hr)
+			if hr.err != nil {
 				return false
 			}
 		} else {
-			hashReader.err = hashReader.StepIn()
-			if hashReader.err != nil {
+			hr.err = hr.StepIn()
+			if hr.err != nil {
 				return false
 			}
 
-			hashReader.err = hashReader.traverse()
-			if hashReader.err != nil {
+			hr.err = hr.traverse()
+			if hr.err != nil {
 				return false
 			}
 
-			hashReader.err = hashReader.StepOut()
-			if hashReader.err != nil {
+			hr.err = hr.StepOut()
+			if hr.err != nil {
 				return false
 			}
 		}
 	}
 
-	next := hashReader.ionReader.Next()
+	next := hr.ionReader.Next()
 	if !next {
-		hashReader.err = hashReader.ionReader.Err()
+		hr.err = hr.ionReader.Err()
 	}
 
-	hashReader.currentType = hashReader.ionReader.Type()
+	hr.currentType = hr.ionReader.Type()
 
 	return next
 }
 
-func (hashReader *hashReader) Err() error {
-	return hashReader.err
+// Err returns an error if a previous call call to Next has failed.
+func (hr *hashReader) Err() error {
+	return hr.err
 }
 
-func (hashReader *hashReader) Type() ion.Type {
-	return hashReader.ionReader.Type()
+// Type returns the type of the Ion value the Reader is currently positioned on.
+// It returns NoType if the Reader is positioned before or after a value.
+func (hr *hashReader) Type() ion.Type {
+	return hr.ionReader.Type()
 }
 
-func (hashReader *hashReader) IsNull() bool {
-	return hashReader.ionReader.IsNull()
+// IsNull returns true if the current value is an explicit null. This may be true
+// even if the Type is not NullType (for example, null.struct has type Struct).
+func (hr *hashReader) IsNull() bool {
+	return hr.ionReader.IsNull()
 }
 
-func (hashReader *hashReader) FieldName() *string {
-	return hashReader.ionReader.FieldName()
+// FieldName returns the field name associated with the current value. It returns
+// nil if there is no current value or the current value has no field name.
```

Comment:
This is the one API that returns a pointer to a value instead of a value. We should state the reason and the cases I think the corresponding interface in ion-go has the text that we can reuse.  https://github.com/amzn/ion-go/blob/de186c33b67694e300a5d190e469f0d7f156a62e/ion/reader.go#L84

Reply:
The text came from the reader API in Ion Go.